x/poolmanager/routes.go

```diff
@@ -0,0 +1,426 @@
+			}
+		}
+	}
+
+	// Set the route map in state
+	// NOTE: This is done with the non map version of the route graph
+	// If we used maps here, the serialization would be non-deterministic
+	osmoutils.MustSet(ctx.KVStore(k.storeKey), types.KeyRouteMap, &routingGraph)
+	return routingGraph, nil
+}
+
+// generateAllDenomPairRoutes generates all possible routes between tokens, without filtering out pools based on liquidity.
+// This returns the pools and route map for use in other functions and does not set the route graph in state.
+func (k Keeper) generateAllDenomPairRoutes(ctx sdk.Context) ([]types.PoolI, types.RoutingGraphMap, error) {
+	// Get all the pools
+	pools, err := k.AllPools(ctx)
+	if err != nil {
+		return []types.PoolI{}, types.RoutingGraphMap{}, err
+	}
+
+	// Create a routingGraph to represent possible routes between tokens
+	var routingGraph types.RoutingGraph
+
+PoolLoop:
+	// Iterate through the pools
+	for _, pool := range pools {
+		// Some of the first cw pools created have a malformed response and are no longer in use. Remove these pools to prevent issues.
+		tokens := pool.GetPoolDenoms(ctx)
+		for _, token := range tokens {
+			if strings.Contains(token, "pool_asset_denoms") {
+				continue PoolLoop
+			}
+		}
+
+		poolID := pool.GetId()
+		// Create edges for all possible combinations of tokens
+		for i := 0; i < len(tokens); i++ {
+			for j := i + 1; j < len(tokens); j++ {
+				// Add edges with the associated token
+				routingGraph.AddEdge(tokens[i], tokens[j], tokens[i], poolID)
+				routingGraph.AddEdge(tokens[j], tokens[i], tokens[j], poolID)
+			}
+		}
+	}
+
+	// Convert the route graph to a map for easier access
+	routeMap := convertToMap(&routingGraph)
+
+	return pools, routeMap, nil
+}
+
+// GetRouteMap returns the route map that is stored in state.
+// It converts the route graph stored in the KVStore to a map for easier access.
+func (k Keeper) GetRouteMap(ctx sdk.Context) (types.RoutingGraphMap, error) {
+	var routeGraph types.RoutingGraph
+
+	found, err := osmoutils.Get(ctx.KVStore(k.storeKey), types.KeyRouteMap, &routeGraph)
+	if err != nil {
+		return types.RoutingGraphMap{}, err
+	}
+	if !found {
+		return types.RoutingGraphMap{}, fmt.Errorf("route map not found")
+	}
+
+	routeMap := convertToMap(&routeGraph)
+
+	return routeMap, nil
+}
+
+// getDirectRouteWithMostLiquidity returns the single hop route with the highest liquidity between an input denom and output denom.
+func (k Keeper) getDirectRouteWithMostLiquidity(ctx sdk.Context, inputDenom, outputDenom string, routeMap types.RoutingGraphMap) (uint64, error) {
+	// Get all direct routes from the input denom to uosmo
+	directRoutes := findRoutes(routeMap, inputDenom, outputDenom, 1)
+
+	// Store liquidity for all direct routes found
+	routeLiquidity := make(map[string]osmomath.Int)
+	for _, route := range directRoutes {
+		liquidity, err := k.getPoolLiquidityOfDenom(ctx, route[0].PoolId, outputDenom)
```

Comment:
These are direct routes. The findRoute method return a [][]Route struct to be compatible with single, double, and triple hop routes. Since we are iterating over direct routes, we are safe to utilize the 0 index since that is all there will be.